### ISSUE TITLE
summary(): For problem CIs, show value reached (instead of NA) with  enhanced warning !!! flag

### ISSUE THREAD

**Issue by tbates** [OpenMx/OpenMx#75]
Stimulated by #74 , can we be more helpful with suggestions for bad CIs? 

`summary` says:

      To investigate missing CIs, run summary() again, with verbose=T, to see CI details. 

That shows failed CIs with alpha level not reached and either `infeasible non-linear constraint` or `iteration limit/blue` as the reason.

Is there a way to request the machine to bang-on for longer with the code blue items?

And for the infeasible items, if the true interval extends beyond the value reached, but constraints pin the value at that point, might it not help the user more to tell them what the value reached was,  prefaced with a ">" or "<" sign in the summary (and still have the !!! warning).

i.e. instead of 

    confidence intervals:
                                 lbound   estimate    ubound note
    top.e_cp[2,2]                  NA       1.0000      NA     !!!

say:

    confidence intervals:
                                 lbound   estimate    ubound    note
    top.e_cp[2,2]                0.9718   1.0000       1.0082    !!!

or even better

    confidence intervals:
                                 lbound   estimate    ubound    note
    top.e_cp[2,2]                0.9718   1.0000       1.0082    !!! alpha level not reached infeasible non-linear constraint

For reference, copied below is the relevant content of  a `$CIdetail` table from a model with bad CIs

```
 x$CIdetail[,c(1:4, 37:38)]
                  parameter      value  side      fit              diagnostic                       statusCode
1             top.a_cp[1,1] -0.6878228 lower 14322.28                 success                               OK
2             top.a_cp[1,1]  0.6878220 upper 14322.28                 success                               OK
3             top.a_cp[3,3] -0.9646925 lower 14322.28                 success                               OK
4             top.a_cp[3,3]  0.9647116 upper 14322.28                 success                               OK
5             top.e_cp[1,1]  0.7258788 lower 14322.28                 success                               OK
6             top.e_cp[1,1]  0.8941798 upper 14322.28                 success                               OK
7             top.e_cp[2,2]  0.9718287 lower 14470.54 alpha level not reached infeasible non-linear constraint
8             top.e_cp[2,2]  1.0082595 upper 14497.86 alpha level not reached infeasible non-linear constraint
9             top.e_cp[3,3]  0.7550760 lower 14448.68 alpha level not reached infeasible non-linear constraint
10            top.e_cp[3,3]  0.9097077 upper 14322.28                 success infeasible non-linear constraint
11 top.cp_loadings_std[1,1]  0.4972017 lower 14322.28                 success infeasible non-linear constraint
12 top.cp_loadings_std[1,1]  0.6108524 upper 14322.28                 success infeasible non-linear constraint
13 top.cp_loadings_std[2,1]  0.8346015 lower 14322.28                 success infeasible non-linear constraint
14 top.cp_loadings_std[2,1]  0.9997913 upper 14325.02 alpha level not reached             iteration limit/blue
15 top.cp_loadings_std[3,1]  0.6268664 lower 14318.44 alpha level not reached             iteration limit/blue
16 top.cp_loadings_std[3,1]  0.6268664 upper 14318.44 alpha level not reached             iteration limit/blue
17 top.cp_loadings_std[4,1]  0.5056475 lower 14318.44 alpha level not reached             iteration limit/blue
18 top.cp_loadings_std[4,1]  0.5056475 upper 14318.44 alpha level not reached             iteration limit/blue
19 top.cp_loadings_std[5,1]  0.5763171 lower 14318.44 alpha level not reached             iteration limit/blue
20 top.cp_loadings_std[5,1]  0.5763171 upper 14318.44 alpha level not reached             iteration limit/blue
21 top.cp_loadings_std[6,1]  0.5799006 lower 14318.44 alpha level not reached             iteration limit/blue
22 top.cp_loadings_std[6,1]  0.5799006 upper 14318.44 alpha level not reached             iteration limit/blue
23 top.cp_loadings_std[1,2]  0.2049466 lower 14318.44 alpha level not reached             iteration limit/blue
24 top.cp_loadings_std[1,2]  0.2049466 upper 14318.44 alpha level not reached             iteration limit/blue
25 top.cp_loadings_std[3,2]  0.2535793 lower 14318.44 alpha level not reached             iteration limit/blue
26 top.cp_loadings_std[3,2]  0.2535793 upper 14318.44 alpha level not reached             iteration limit/blue
27 top.cp_loadings_std[4,2]  0.4057909 lower 14318.44 alpha level not reached             iteration limit/blue
28 top.cp_loadings_std[4,2]  0.4057909 upper 14318.44 alpha level not reached             iteration limit/blue
29 top.cp_loadings_std[5,2]  0.8172261 lower 14318.44 alpha level not reached             iteration limit/blue
30 top.cp_loadings_std[5,2]  0.8172261 upper 14318.44 alpha level not reached             iteration limit/blue
31 top.cp_loadings_std[6,2]  0.2704143 lower 14318.44 alpha level not reached             iteration limit/blue
32 top.cp_loadings_std[6,2]  0.2704143 upper 14318.44 alpha level not reached             iteration limit/blue
33 top.cp_loadings_std[3,3]  0.5236328 lower 14318.44 alpha level not reached             iteration limit/blue
34 top.cp_loadings_std[3,3]  0.5236328 upper 14318.44 alpha level not reached             iteration limit/blue
35 top.cp_loadings_std[4,3]  0.2304582 lower 14318.44 alpha level not reached             iteration limit/blue
36 top.cp_loadings_std[4,3]  0.2304582 upper 14318.44 alpha level not reached             iteration limit/blue
37 top.cp_loadings_std[6,3]  0.1536341 lower 14318.44 alpha level not reached             iteration limit/blue
38 top.cp_loadings_std[6,3]  0.1536341 upper 14318.44 alpha level not reached             iteration limit/blue
39          top.as_std[1,1] -0.3579901 lower 14318.44 alpha level not reached             iteration limit/blue
40          top.as_std[1,1] -0.3579901 upper 14318.44 alpha level not reached             iteration limit/blue
41          top.es_std[1,1]  0.7206583 lower 14318.44 alpha level not reached             iteration limit/blue
42          top.es_std[1,1]  0.7206583 upper 14318.44 alpha level not reached             iteration limit/blue
43          top.es_std[2,2]  0.4429754 lower 14318.44 alpha level not reached             iteration limit/blue
44          top.es_std[2,2]  0.4429754 upper 14318.44 alpha level not reached             iteration limit/blue
45          top.es_std[3,3]  0.5182130 lower 14318.44 alpha level not reached             iteration limit/blue
46          top.es_std[3,3]  0.5182130 upper 14318.44 alpha level not reached             iteration limit/blue
47          top.es_std[4,4]  0.7256331 lower 14318.44 alpha level not reached             iteration limit/blue
48          top.es_std[4,4]  0.7256331 upper 14318.44 alpha level not reached             iteration limit/blue
49          top.es_std[6,6]  0.7529860 lower 14318.44 alpha level not reached             iteration limit/blue
50          top.es_std[6,6]  0.7529860 upper 14318.44 alpha level not reached             iteration limit/blue
> 
```

**Comment by jpritikin** [comment]
The summary output puts both the upper and lower bound on the same line,

> confidence intervals:
>                              lbound   estimate    ubound    note
> top.e_cp[2,2]                0.9718   1.0000       1.0082    !!! alpha level not reached infeasible non-linear constraint

That makes it awkward to report the diagnostics---you'd need 2 copies of them, one for the lbound and one for the ubound.

> And for the infeasible items, if the true interval extends beyond the value reached, but constraints pin the value at that point, might it not help the user more to tell them what the value reached was, prefaced with a ">" or "<" sign in the summary (and still have the !!! warning).

When optimization fails, I don't think we know how to interpret the results.

> Is there a way to request the machine to bang-on for longer with the code blue items?

Yes, the simple way is to set the global option:

**Comment by tbates** [comment]
Should we have mxTryHardCI which would increase major and rerub the code blues?

**Comment by tbates** [comment]
Re the pivot-table issue of reporting problems for lower and upper, perhaps !!! When anything's wrong, and " lower fine, upper bad: heres why" in the description string

**Comment by jpritikin** [comment]
> Should we have mxTryHardCI which would increase major and re-run the code blues?

Seems like a feature that could be added to `omxRunCI`

**Comment by RMKirkpatrick** [comment]
I agree with Tim that the CI details table is hard to read.  But I agree with Joshua that the main CI table would be harder to read if the diagnostic information for both confidence limits were on the same line.

> Should we have mxTryHardCI which would increase major and re-run the code blues?

I agree with Joshua: that's the sort of thing that would fit well into `omxRunCI()`.  BTW, if OpenMx is using the constrained representation of the confidence-limit optimization problem, then it automatically re-tries failed confidence limits as many as 3 times, with random perturbation of start values between attempts.

> And for the infeasible items, if the true interval extends beyond the value reached, but constraints pin the value at that point, might it not help the user more to tell them what the value reached was, prefaced with a ">" or "<" sign in the summary (and still have the !!! warning).

When OpenMx is using the constrained representation of the confidence-limit optimization problem, the "infeasible non-linear constraint" is often going to refer to the inequality constraint that keeps the worsening in -2logL greater than the chi-square critical value.  If that constraint can't be satisfied, the value obtained for the confidence limit will almost always fail to achieve the target alpha level.

I recall discussion in the past about whether or not suspect confidence limits should be reported as `NA` in the main CI table.  I believe we decided that they should, because it sends a much stronger hint to the user that something _may_ be wrong with them.  That is, we were concerned that users would ignore the "!!!" alert if they also got seemingly valid numerical results in the same table.

**Comment by mcneale** [comment]
On Apr 18, 2018, at 10:36 AM, RMKirkpatrick <notifications@github.com<mailto:notifications@github.com>> wrote:


I agree with Tim that the CI details table is hard to read. But I agree with Joshua that the main CI table would be harder to read if the diagnostic information for both confidence limits were on the same line.

Should we have mxTryHardCI which would increase major and re-run the code blues?

I agree with Joshua: that's the sort of thing that would fit well into omxRunCI(). BTW, if OpenMx is using the constrained representation of the confidence-limit optimization problem, then it automatically re-tries failed confidence limits as many as 3 times, with random perturbation of start values between attempts.

And for the infeasible items, if the true interval extends beyond the value reached, but constraints pin the value at that point, might it not help the user more to tell them what the value reached was, prefaced with a ">" or "<" sign in the summary (and still have the !!! warning).

When OpenMx is using the constrained representation of the confidence-limit optimization problem, the "infeasible non-linear constraint" is often going to refer to the inequality constraint that keeps the worsening in -2logL greater than the chi-square critical value. If that constraint can't be satisfied, the value obtained for the confidence limit will almost always fail to achieve the target alpha level.

I recall discussion in the past about whether or not suspect confidence limits should be reported as NA in the main CI table. I believe we decided that they should, because it sends a much stronger hint to the user that something may be wrong with them. That is, we were concerned that users would ignore the "!!!" alert if they also got seemingly valid numerical results in the same table.

I don’t agree with this behavior.  It seems to me that, e.g., a^2 lower limit of 0 would be suppressed if that value did not generate a -2lnL difference of 3.84 when a^2 is fixed to zero and all other parameters are free.  I agree that we should let users know that a boundary has effectively been encountered, but it is fine to report that this occurred when the estimate zero.   Saying we have no idea what it is, when in fact we do (it’s less than zero but the optimizer can’t get there), seems an unwarranted suppression of information.

**Comment by jpritikin** [comment]
Rob's comment was much more general than yours Mike. If there are specific, narrow situations where we might report the upper/lower bound even when optimization failed then exactly what are these specific situations?

**Comment by mcneale** [comment]
I understand, I was giving an example.  I think there are three main situations where the CI won't be at the 'right' place.  
1. Optimization failure - code 4 or 5 or something pretty toxic.  NA fine here
2. A free parameter hits explicit bound.  Report bound but note that 3.84 threshold possibly not met.
3. An algebra element hits an implicit bound, due to non-linear constraints or implicit bound such as a^2 not possible to be < zero.  Report what was found but again issue a warning.

**Comment by tbates** [comment]
As I work with this, it seems _more_ misleading to print NA when we actually have a value, but consider it suspect.

In the models I'm trying to report (`$CIdetails` copied in issue above), all the NAs are either constraints, where the constrained value makes sense to the user if she sees value (rather than NA). (cases 2 and 3 from @mcneale's post above this one). 

The rest (code blues) seem to be cases where neither lower nor upper bound for an algebra cell has moved from the estimate of the parameter. They're cells where the underlying matrices fix the algebra value.

Rather than printing NA, it would help the user to detect this, report the value, and, in the notes say "**!!! Algebra may be fixed at this cell**" or similar.

**Comment by tbates** [comment]
Here's an example called `cp3` (A three common factor CP twin model with various bad CIs)
it's named "cp3.rda.txt" for upload, but should `load()` fine.

    load("https://github.com/OpenMx/OpenMx/files/1924799/cp3.rda.txt")

[cp3.rda.txt](https://github.com/OpenMx/OpenMx/files/1924799/cp3.rda.txt)

**Comment by RMKirkpatrick** [comment]
Somewhat off-topic, but: Tim, you might want to give Nelder-Mead a try.  I created a custom compute plan to calculate the CIs in your 'cp3' model with Nelder-Mead using the following syntax:
```
plan3 <- omxDefaultComputePlan(intervals=TRUE)
plan3 <- mxComputeSequence(steps=list(CI=plan3$steps$CI))
plan3$steps$CI$constraintType <- "ineq"
plan3$steps$CI$plan <- mxComputeNelderMead(ineqConstraintMthd="eqMthd",eqConstraintMthd="backtrack",centerIniSimplex=F)
cp3_3 <- mxModel(cp3,plan3)
cp3_3fit <- mxRun(cp3_3)
```
The results (attached) are surprisingly good!

[180418--cp3_3fit_CIdetail.csv.txt](https://github.com/OpenMx/OpenMx/files/1925154/180418--cp3_3fit_CIdetail.csv.txt)

**Comment by mcneale** [comment]
Oddly, loading from the link doesn't work, but downloading the file first does.
``
> load("https://github.com/OpenMx/OpenMx/files/1924799/cp3.rda.txt")
Error in readChar(con, 5L, useBytes = TRUE) : cannot open the connection
In addition: Warning message:
In readChar(con, 5L, useBytes = TRUE) :
  cannot open compressed file 'https://github.com/OpenMx/OpenMx/files/1924799/cp3.rda.txt', probable reason 'No such file or directory'
> 
> load("~/Downloads/cp3.rda.txt")
``

**Comment by tbates** [comment]
Sounds like  `omxRunCI` could usefully gain two new options:
* tryHard (default = TRUE). Re-run CIs which failed code blue, using a custom compute plan that ups major iterations
* method one of c("Nelder-Mead", ...)

**Comment by RMKirkpatrick** [comment]
> tryHard (default = TRUE). Re-run CIs which failed code blue, using a custom compute plan that ups major iterations

Is the idea that the user will have already tried running confidence intervals, and this feature would only re-run the subset of confidence limits that timed out with status Blue?  I'm a bit skeptical that confidence limits that end up with status Blue are going to converge with extra iterations. Maybe the function should just have a `maxIter` argument instead.

> method one of c("Nelder-Mead", ...)

`omxRunCI()` (and `omxParallelCI()`) already has an `optimizer` argument, and could be modified to accept "Nelder-Mead" as a value for that argument.  Also, there are multiple ways Nelder-Mead could conceivably be used for CI optimization.  For instance, if 'cp3' had equality MxConstraints in it, I would not expect `plan3` to be a good compute plan, because `eqConstraintMthd="backtrack"` is not the best method for handling equality constraints.

**Comment by RMKirkpatrick** [comment]
Anyhow, for reference, attached here is some syntax that produces four compute plans that, to me, seem reasonably likely to work well at estimating confidence intervals with Nelder-Mead.
[180419--NelderMeadCIs.R.txt](https://github.com/OpenMx/OpenMx/files/1928392/180419--NelderMeadCIs.R.txt)

**Comment by tbates** [comment]
Just FYI, I tried the Nelder-Mead here, but it failed to run the final version of our model... (error pasted below)

For the final model, with omxRunCI _every_ CI of interest is NA and the values in details are just the start values - claims to have had code blue without moving the value at all...  :-(

`tryHard` with all CIs being run each time is likely to be a long time coming :-(

Need some automated way to run each CI, one at a time, behaving intelligently when a given CI fails.


```
Running 3factorCPfinal with 31 parameters
Error: The job for model '3factorCPfinal' exited abnormally with the error message: initial simplex is not feasible; specify it differently, try different start values, or use mxTryHard()
ited abnormally with the error message: initial simplex is not feasible; specify it differently, try different start values, or use mxTryHard()
> 
```

**Comment by mcneale** [comment]
Not that I’m aware of.  As always, providing a full running script that illustrates the issue helps development team to investigate.

**Comment by tbates** [comment]
Yes, apologies. I've been working on `umxConfint` to try and make working with CIs easy for users, and to populate diagrams with CIs so it benefits everyone. But here's plain OpenMx code.

First, you need the final reduced model

[reduced_model_cp3h.rda.txt](https://github.com/OpenMx/OpenMx/files/1928954/reduced_model_cp3h.rda.txt)

cmd-click the link to download. Set suffix to rda, and drop file onto R.app. It should load into workspace as object "cp3h").

First, we just ask for 2 CIs
```Splus
just2 = c("top.cp_loadings_std[1,1]", "top.cp_loadings_std[3,1]")
good = omxRunCI(mxModel(cp3h, mxCI(just2)))
summary(good)$CI
                            lbound  estimate    ubound note
top.cp_loadings_std[1,1] 0.4972007 0.5572190 0.6108524     
top.cp_loadings_std[3,1] 0.5632765 0.6268682 0.6842034
```

And now a whole bunch (all valid).

```Splus
all = c("top.a_cp[1,1]", "top.a_cp[3,3]", "top.e_cp[1,1]", "top.e_cp[2,2]", "top.e_cp[3,3]", "top.cp_loadings_std[1,1]", "top.cp_loadings_std[2,1]", "top.cp_loadings_std[3,1]", "top.cp_loadings_std[4,1]", "top.cp_loadings_std[5,1]", "top.cp_loadings_std[6,1]", "top.cp_loadings_std[1,2]", "top.cp_loadings_std[3,2]", "top.cp_loadings_std[4,2]", "top.cp_loadings_std[5,2]", "top.cp_loadings_std[6,2]", "top.cp_loadings_std[3,3]", "top.cp_loadings_std[4,3]", "top.cp_loadings_std[6,3]", "top.as_std[1,1]", "top.es_std[1,1]", "top.es_std[2,2]", "top.es_std[3,3]", "top.es_std[4,4]", "top.es_std[6,6]")

bad = omxRunCI(mxModel(cp3h, mxCI(all)))
summary(tmp2)$CI
                             lbound   estimate    ubound note
top.a_cp[1,1]            -0.6878228  0.5850266 0.6878220     
top.a_cp[3,3]            -0.9646915 -0.7164735 0.9647084     
top.e_cp[1,1]             0.7258788  0.8110141 0.8942316     
top.e_cp[2,2]                    NA  1.0000000 1.0000000  !!!
top.e_cp[3,3]                    NA  0.6976143 0.9097068  !!!
top.cp_loadings_std[1,1]  0.4972007  0.5572190 0.6108524     
top.cp_loadings_std[2,1]  0.8346015  0.8965314        NA  !!!
top.cp_loadings_std[3,1]         NA  0.6268682        NA  !!! <- look here!!!
top.cp_loadings_std[4,1]         NA  0.5056488        NA  !!!
top.cp_loadings_std[5,1]         NA  0.5763183        NA  !!!
top.cp_loadings_std[6,1]         NA  0.5799022        NA  !!!
top.cp_loadings_std[1,2]         NA  0.2049450        NA  !!!
top.cp_loadings_std[3,2]         NA  0.2535771        NA  !!!
top.cp_loadings_std[4,2]         NA  0.4057896        NA  !!!
top.cp_loadings_std[5,2]         NA  0.8172253        NA  !!!
top.cp_loadings_std[6,2]         NA  0.2704125        NA  !!!
top.cp_loadings_std[3,3]         NA  0.5236327        NA  !!!
top.cp_loadings_std[4,3]         NA  0.2304570        NA  !!!
top.cp_loadings_std[6,3]         NA  0.1536319        NA  !!!
top.as_std[1,1]                  NA -0.3579851        NA  !!!
top.es_std[1,1]                  NA  0.7206602        NA  !!!
top.es_std[2,2]                  NA  0.4429802        NA  !!!
top.es_std[3,3]                  NA  0.5182121        NA  !!!
top.es_std[4,4]                  NA  0.7256333        NA  !!!
top.es_std[6,6]                  NA  0.7529859        NA  !!!
```

See the CI marked "look here"? It fails (but worked above when it was one of just 2 requested.

**Comment by mcneale** [comment]
My initial thought is that the parameter estimates might not be reset to their MLE's when moving from estimating one CI to the next.  That would give different starting values, possibly really bad ones, when computing a series of CIs.  This hypothesis should be easy to test and if it is the case, I suggest that the behavior be changed to include the reset.

**Comment by RMKirkpatrick** [comment]
I'm pretty sure I know what's happening here.  By default, `omxRunCI()` uses SLSQP; when SLSQP is in use, the default behavior is to use the inequality-constrained representation; when the inequality-constrained representation is in use, the CI step in the default compute plan encases its optimization step inside another step that causes failed confidence limits to be retried as many as 3 times, _with start values randomly perturbed between attempts_.  In other words, the default behavior of `omxRunCI()` is partly stochastic.

**Comment by mcneale** [comment]
RMK's answer does not explain why the same CI works in one list but not another.  If the parameters are not reset to their MLEs after each interval optimization (upper or lower), then a cascade of failures such as described by tbates would ensue.  Sanitizing the starting values every time would seem to be the solution, no?

**Comment by RMKirkpatrick** [comment]
> RMK's answer does not explain why the same CI works in one list but not another.

I think it _could_ explain it based on what I knew when I posted it, but now that I have more information, I don't think so anymore.  The following syntax removes the stochastic element from the MxComputeConfidenceInterval step:
```
library(umx)
load("/home/rmk/OpenMx_dev/reduced_model_cp3h.rda")
mxOption(NULL,"Default optimizer","SLSQP")

ciOpt <- mxComputeGradientDescent(engine="SLSQP",nudgeZeroStarts=FALSE)
pciplan <- mxComputeSequence(
	steps=list(
		CI=mxComputeConfidenceInterval(
			constraintType="ineq",
			plan=ciOpt
		)))
just2 = c("top.cp_loadings_std[1,1]", "top.cp_loadings_std[3,1]")
good = mxRun(mxModel(cp3h, mxCI(just2), pciplan))
summary(good)$CI

                            lbound  estimate    ubound note
top.cp_loadings_std[1,1] 0.4972007 0.5572190 0.6108524     
top.cp_loadings_std[3,1] 0.5632765 0.6268682 0.6842034     

all = c(
	"top.a_cp[1,1]", "top.a_cp[3,3]", "top.e_cp[1,1]", "top.e_cp[2,2]", "top.e_cp[3,3]", "top.cp_loadings_std[1,1]", 
	"top.cp_loadings_std[2,1]", "top.cp_loadings_std[3,1]", "top.cp_loadings_std[4,1]", "top.cp_loadings_std[5,1]", "top.cp_loadings_std[6,1]", 
	"top.cp_loadings_std[1,2]", "top.cp_loadings_std[3,2]", "top.cp_loadings_std[4,2]", "top.cp_loadings_std[5,2]", "top.cp_loadings_std[6,2]", 
	"top.cp_loadings_std[3,3]", "top.cp_loadings_std[4,3]", "top.cp_loadings_std[6,3]", "top.as_std[1,1]", "top.es_std[1,1]", "top.es_std[2,2]", 
	"top.es_std[3,3]", "top.es_std[4,4]", "top.es_std[6,6]")
bad = mxRun(mxModel(cp3h, mxCI(all), pciplan))
summary(bad)$CI

                             lbound   estimate    ubound note
top.a_cp[1,1]            -0.6878228  0.5850266 0.6878220     
top.a_cp[3,3]            -0.9646915 -0.7164735 0.9647082     
top.e_cp[1,1]             0.7258788  0.8110141 0.8922750     
top.e_cp[2,2]                    NA  1.0000000        NA  !!!
top.e_cp[3,3]                    NA  0.6976143 0.9097067  !!!
top.cp_loadings_std[1,1]  0.4972007  0.5572190 0.6108524     
top.cp_loadings_std[2,1]  0.8346015  0.8965314 0.9722005     
top.cp_loadings_std[3,1]  0.5632765  0.6268682        NA  !!!
top.cp_loadings_std[4,1]         NA  0.5056488        NA  !!!
top.cp_loadings_std[5,1]         NA  0.5763183        NA  !!!
top.cp_loadings_std[6,1]         NA  0.5799022        NA  !!!
top.cp_loadings_std[1,2]         NA  0.2049450        NA  !!!
top.cp_loadings_std[3,2]         NA  0.2535771        NA  !!!
top.cp_loadings_std[4,2]         NA  0.4057896        NA  !!!
top.cp_loadings_std[5,2]         NA  0.8172253        NA  !!!
top.cp_loadings_std[6,2]         NA  0.2704125        NA  !!!
top.cp_loadings_std[3,3]         NA  0.5236327        NA  !!!
top.cp_loadings_std[4,3]         NA  0.2304570        NA  !!!
top.cp_loadings_std[6,3]         NA  0.1536319        NA  !!!
top.as_std[1,1]                  NA -0.3579851        NA  !!!
top.es_std[1,1]                  NA  0.7206602        NA  !!!
top.es_std[2,2]                  NA  0.4429802        NA  !!!
top.es_std[3,3]                  NA  0.5182121        NA  !!!
top.es_std[4,4]                  NA  0.7256333        NA  !!!
top.es_std[6,6]                  NA  0.7529859        NA  !!!
```
So the upper limit for top.cp_loadings_std[3,1] succeeded the first time, but not the second time...

**Comment by tbates** [comment]
@mcneale yes, I think this is a bug in `omxParallelCI`

Shifting to plain mxRun I get all the CIs (yay)

```Splus
may = mxRun(mxModel(cp3h, mxCI(all)), intervals=T)
Running 3factorCPfinal with 31 parameters                                  
```

Nasty time penalty going back to mxRun, so it will be great to get omxParallelCI working properly.

PS: I saw that CIs are not using parallel. So opening a new issue (#77)

```Splus
umx_time(may)
3factorCPfinal: 40.20 seconds

umx_time(bad)
3factorCPfinal: 12.38 seconds

```

**Comment by RMKirkpatrick** [comment]
Maybe it is a bug in `omxParallelCI()`.  Do you notice any differences between the compute plan in `may$compute` and `bad$compute`.

> Nasty time penalty going back to mxRun, so it will be great to get omxParallelCI working properly.

I think the time penalty reflects whatever additional work the backend had to do to get the CIs right for `may`.  As to why it worked for `may` and not `bad`, my suspicion is that it has to do with some difference in the compute plans that are constructed by `mxRun()` versus `omxParallelCI()`.

**Comment by mcneale** [comment]
@tbates I think the time penalty may not concern parallel, but that the optimizer throws in the towel rather more quickly when the starting values cause an NA snafu.

**Comment by tbates** [comment]
I wonder (given above edits localizing this problem to the optimizer, not differences in `mxRun` or `omxRunCI`) if we would be 
1. Better off leaving the optimizer parameter of `omxRunCI` = NULL (use the current default). Unless SLSQP is nearly always better than the user's chosen optimizer for their particular problem, then, certainly for me, swap optimizers risks blindsiding the user in those cases where they really need our help?

2. Check the CIs computed by `omxRunCI` , and if they contain any notes, print a warning along the lines of:
> One or more CIs might not have run properly. Please inspect these carefully. You might want to switch  optimizers and try again (you were using",  omxQuote(current), ", but alternatives omxQuote(valid opt), " might be better?

Perhaps suggest in the help that SLSQP is better for some problems, and specify which these are (if we know)?

**Comment by jpritikin** [comment]
@tbates At least I fixed one bug, v2.9.6-61-g00be5305b. The major iteration limit was not be accounted for correctly (it was per-mxRun, not per optimization). So the error about iteration limit was precisely accurate.

**Comment by tbates** [comment]
That’s big, no? So it simply ran out of runs after a few CIs...

**Comment by mcneale** [comment]
Does your problem run ok now then, Tim?

**Comment by RMKirkpatrick** [comment]
It occurs to me that using `omxParallelCI()`, with `independentSubmodels=TRUE`, would serve as a workaround for the bug that causes confidence limits later in the list to run out of iterations.

Of course, the bug is patched in the repository now, isn't it?

**Comment by jpritikin** [comment]
> Of course, the bug is patched in the repository now, isn't it?

Yes

**Comment by tbates** [comment]
This now computes the CIs equally well across optimizers and no longer has a fail point of iterations being used up cumulatively instead of refreshing with each CI.

**Comment by mcneale** [comment]
I would add that we could use the output column containing !!! notes differently.  We might use !1 to index an informative note at the end of summary.  Add notes only for the particular !1, !2, !3 etc notes flagged in the note column.  That would possibly save the user from having to mxSummary(model, verbose=T) to see what's up.

**Comment by mcneale** [comment]
Woot!  Thanks for figuring out and fixing the bug, Joshua!  I note I was wrong about the reset of parameter estimates to their MLE's - that is done, but the reset of number of iterations was not until the fix.

**Comment by tbates** [comment]
This issue morphed into fixing a bug for iterations accumulating over multiple CIs and discussion about Nelder Mead. Might be more easily addressed in #78, which references this issue.
